### ISSUE TITLE
malloc error checking

### ISSUE THREAD

**Issue by laochailan** [taisei-project/taisei#47]
As you may know, taisei doesn’t take checking the return value of malloc for NULL seriously and would segfault if NULL is returned.

Now with all the dialog graphics and possibly BGM in the future, it is possible that some people will try to run this on their SoC/toaster and it will run out of memory and maybe return NULL on some kind of system and the user will not know it’s because of that.
Maybe we should define some malloc function that will panic and print an error message if that happens and replace all mallocs by that.

I do not know if the builtin references and linked-list implementations (or anything else) rely on malloc(0) == NULL though. But that could be retained by something like

```
void *tmalloc(size_t size) {
    void *p = malloc(size);
    if(p == 0 && size != 0)
        oh_no();
    return p;
}
```

**Comment by Akaricchi** [comment]
```
 % free -m
              total        used        free      shared  buff/cache   available
Mem:          32087        6092        8298         184       17697       25357
Swap:             0           0           0
```

Seriously though, how do you even recover from a failing malloc, other than directing the player to downloadmoreram.com before exiting? I mean, it seems like too much effort just for the sake of making it crash gracefully. Plus SDL and probably some other libraries that we use invoke malloc internally, and I don't think you can intercept that in a portable way.

**Comment by laochailan** [comment]
I don’t want to exit gracefully. I just want to print a message like “ran out of memory” before exiting instead of having a segfault which could also be caused by other things.

oh_no() would be basically errx().

**Comment by lachs0r** [comment]
Well, IMO the BGM should just be streamed instead of decoding it all at once. But that’s a separate issue.

As for malloc checks, perhaps we should use [mpv’s Tree Allocator](https://github.com/mpv-player/mpv/tree/master/ta)?

**Comment by laochailan** [comment]
Hi lachs0r! :)

Yes, streaming would be better I think.

That tree allocator looks nice, but complicated. It would probably be a lot of work to make it interface with the resource allocation and linked list voodoo without unknowingly breaking something.

**Comment by lachs0r** [comment]
That sounds like a refactoring target, but we could replace all the malloc calls and use it without the tree features where it’s not feasible yet. That still gives us error checking and its debugging features (which do look somewhat useful).

**Comment by Akaricchi** [comment]
I agree about BGM streaming, and still think that running out of RAM is virtually a non-issue.

I get it that you want to give a meaningful message to the user (assuming they even look at the stderr). But this case is one of the easiest to rule out when troubleshooting/debugging, and in my opinion, this is going to be such an extremely rare occurence that if someone does get a segfault, in all likelihood one of us just fucked up and dereferenced a null pointer somewhere.

That said, I'm not opposed to the idea, especially if you can find a good way to deal with in-library allocations. I'd just give it the lowest priority out of any issues I can think of.

**Comment by laochailan** [comment]
I agree with it being low priority.

I think all the libraries we are using are checking their malloc pointers properly. At least [SDL](https://github.com/SDL-mirror/SDL/search?utf8=%E2%9C%93&q=%22SDL_malloc%22&type=Code)
 and [libpng](https://github.com/glennrp/libpng/search?utf8=%E2%9C%93&q=malloc) seem to.

**Comment by makise-homura** [comment]
Well, I have such a habit to NULL-check every thing which ever may be NULL (pointers returned by fopens, mallocs, strdups, etc) just to produce suitable message to user rather than segfault which have do be debugged with gdb, and you may figure out which was the cause only after seeing a call stack. It slightly enlarges the code, but guarantees that user will know the reason (yes, even in nearly-impossible situations as there was enough RAM to load megabytes of graphics and music, but no place to reallocate a string of couple of bytes). It also helps debugging memory leaks and other stuff.
So, looks like a good idea for improvement, but yes, not in a priority.

> IMO the BGM should just be streamed instead of decoding it all at once.
Yes, it's the most adequate way how to implement BGM support (as when I did it, I just tried to make it at least working with libraries already used in project). Now, as there are such things as migrating from SDL to SDL2, it looks like good time to rewrite this using something more suitable.

**Comment by Akaricchi** [comment]
We could probably use SDL_mixer to handle both sound effects and music (which it can stream). It can also load OGGs transparently, if it was built with vorbis support. We can get rid of the OpenAL and FreeALUT dependencies this way, too.

**Comment by laochailan** [comment]
Yes, it might be good to use a library that can do streaming. But back in the day we had some reason to use OpenAL instead of mixer. I just don’t remember what.

SDL2_mixer is a separate dependency though and on most systems it is built with a bunch of other libraries. So in terms of things the user has to install, it will be not much of a gain in that way.

**Comment by makise-homura** [comment]
As I think, if Taisei requires SDL2 and SDL2_ttf, it will not be strange if it will require SDL2_mixer in addition. But, maybe we can let user choose?

If we'll have standartized interface in audio.h and bgm.h, why not have audio_alut.c and bgm_alut.c, audio_sdl2m.c and bgm_sdl2m.c, and nosound.c with stubs for those who don't want sound / don't have an ability to install OpenAL+ALUT or SDL2_mixer? (If the one has either OpenAL+ALUT and SDL2_mixer installed, we may let him at least ability to choose which facility to use).

As far as I know, gzdoom has similar thing about music: it can play it through FMOD or FluidSynth, and it may be configured in cmake command line.

(Eh, looks like good idea for me to work on something next days, as I'm deeply familiar with audio code of Taisei for now. :)

**Comment by laochailan** [comment]
I don’t think we should go out of our way to support multiple different backends. In the end, openal and SDL are probably doing the same thing to the hardware.

I think using SDL2_mixer for sound and streaming would be sufficient.

But we could use your help also in other areas. #52 is also kind of sound related and could make you more familiar with the ingame stuff.

**Comment by makise-homura** [comment]
Ok, maybe, simply migrating to SDL2_mixer can be the most obvious thing to do both with music streaming and simultaneous sound stuff related to #52, if there are no significant worries about that forgotten reason to choose OpenAL instead of it.

**Comment by laochailan** [comment]
We went over it again on irc and the reason was 3D sound I think and we probably wouldn’t need that for this game.

**Comment by Akaricchi** [comment]
I'm closing this as wontfix for two reasons:

1. This is a very low priority issue.
2. There is not much you can do in case of a failing malloc anyway. Taisei is not the kind of program that can recover from an allocation failure. The only behavior that makes sense (for us) is to call `abort()`. You can't even count on printing an error message, because it's likely that this very action is going to try and fail to allocate memory somewhere down the call stack. And that's assuming that the OS doesn't [overcommit](https://www.etalabs.net/overcommit.html), which Linux does by default, by the way.

**Comment by j123123** [comment]
JFYI: https://www.viva64.com/ru/b/0558/ https://www.viva64.com/en/b/0558/

**Comment by j123123** [comment]
Quote:
>At the beginning of the address space, one or more pages of memory are protected by the operating system. This allows to identify errors of addressing to memory by a null pointer, or the pointer of value close to 0.
>
>In various operating systems, different amounts of memory are reserved for these purposes. Besides, in some operating systems this value is configurable. Therefore, it makes no sense to call a specific number of bytes of memory reserved. But in order to direct the reader, I would say that in Linux systems the standard value is 64Kb.
>
>It is important that, adding any sufficiently large number to a null pointer, you can "strike out" the control memory pages and accidentally get into any unprotected page records. Thus, it is possible to corrupt somewhere some information, but the operating system will not notice and generate any signal/exception.

So it's actually better to make some xmalloc(), xrealloc(), xstrdup() etc function wrapprer, which call exit(ENOMEM) when failed instead hope that you don't
> "strike out" the control memory pages and accidentally get into any unprotected page records

**Comment by Akaricchi** [comment]
>So it's actually better to make some xmalloc(), xrealloc(), xstrdup() etc function wrapprer, which call exit(ENOMEM) when failed instead hope that you don't

I'd be ok with that; patches welcome :)

Not directly related, but stealing [mpv's ta](https://github.com/mpv-player/mpv/tree/master/ta) and refactoring around it, as @lachs0r suggested earlier, might also be a good idea to clean up some allocation-heavy code.
